inst/training_202307/202307.qmd

```diff
@@ -429,16 +484,12 @@ None of the test result is _stable_:
 
     This is the most instable environment changed by pushes to any of the dependencies. Also: the main branch is usually not guaranteed to be correct and you might pull broken package version.
 
-# A few caveats:
-
-What it is not about:
-
-- This does not perform matrix testing of _all_ dependent package versions.
-- This only looks into package versions - in particular: it does not check R versions or system architectures.
+Therefore, this test should be triggered periodically and not on code change (unless changes includes dependencies). Triggering it on change might give a false message that a failure stems from code changes.
 
 # Future work:
 
```

Comment:
Binary search rather than walk through all versions of direct deps during min search.

**even better version**: binary search by dates and just use exact snapshots for those dates. This will get you very close to the edge you're searching for, even if it doesn't exactly find it.

Reply:
Not only does it get you close to what wha concurrent_min would do, it arguably gives end users of packages something more important, an actual real cran state within which the package would have worked.

A bit of a caveat. could get unlucky with a situation like 

- v 0.1.0 of dep works
- v 0.1.1 breaks our pkg
- v 0.1.2 of dep works

But there are ways of detecting that for slight extra cost

Reply:
ie when your binary search finds an edge, test the versions directly on either side to detect this.

Reply:
> Interesting. But note one difference. This algorithm will output you env (which then you can translate into DESC specs) whereas my goal is (at least it was at the very beginning) to take specs from DESC and return true/false. You have described a tool to give you correct answer but here I more look for an answer validator. The intention is that if on false, pkg maintainer will correct it to be true. In particular: it doesn't need to be the "true" min specs - just the specs that doesn't fail. Nevertheless, it's quite interesting. Thanks for sharing. But I see it as an independent product.

I don't understand the reasoning here. Your min plan has to do a search through the possible "environments", as you're calling them, checking them until you get to one that succeeds. 

What I describe above is the same, but a more efficient search which should have fewer failed  checks (substantially so for packages with many versions, O(n) vs O(log n) if we were searching among pkg versions) before returning. It deals in environments only exactly the same way the min version detection algorithm already does. 

The trick here will be to do the fast fourier transform approach. There are thousands of dates you might consider, but we only have to include ones that could be change points, ie ones at which at least one of the reverse dependencies got a new version. That will reduce the thousands of potential dates to a few dozen (low hundreds at most I'd expect), and then the binary search will clobber that down to a few checks total (200 dates would translate to 8 cohort checks, or up to ~16 if you do the double checking for point breakages).

Reply:
> I don't understand the reasoning here. Your min plan has to do a search through the possible "environments", as you're calling them, checking them until you get to one that succeeds.

That is partially true. The main goal for min deps check is to check whether everything works if I use minimal dependencies as per DESCRIPTION file. In particular: I am not looking for (searching) the optimum minimal deps possible. And I am not checking whether min deps in DESC is the optimum one.
Let me put that in the example. Let's say the pkg is fully compatible with all versions of dplyr starting from 1.0.0 but there is "dplyr >= 1.2.3" for some reason. For the check I am thinking of, it's a success as tested pkg does not return any failure. Note that I am not failing because v1.2.3 is different than the _true_ minimal v1.0.0.

Let's then dive into "(...) as per DESCRIPTION file" part of the above sentence. As you correctly pointed out earlier, dependency specifications cannot be really taken as is (to put it simply: replace ">=" with "==") and most likely this would lead to the joint dependency tree conflicts. Please refer to the dplyr >= 1.0.0 together with tibble >= 1.0.0 example discussed earlier. This has to be "corrected" - I can either fail (which means to require inputs that would make joint tree to be resolved) or I can "correct" it on-the-fly. As you correctly pointed out earlier - the former will break encapsulation in a really bad way so I gave a try with the latter and achieve it with PPM snapshot. Now when it comes to the PPM, there is actually no search at all. I take release date for each direct dependent pkgs, calculate max out of that collection and use that date as PPM snapshot. For the min_isloated, it's very similar - take the release date for each directly dependent pkgs, resolve itself (that's the difference) and then combine resolutions and resolve once again. No search / optimization included.

To use mathematical analogy where foo represents a checking function that returns 1 on success and 0 on failure. I can see that you are thinking about this problem as **find** (!) min _x_ for which f(x) > 0. I see it more simpler **as for a given** (!) _x_ check if f(x) > 0. Not so rarely f(x) cannot be calculated and I needed to f(g(x)) > 0. The g function can be debatable but still we have x already given.

Finding an optimal min deps is a definitely interesting question and I very much like your idea how to do that but this does not really belong into "check" role of this package. IMHO it should be a new product.